src/OpenTelemetry/Trace/Span.cs

```diff
@@ -328,6 +332,12 @@ public void End()
 
         public void End(DateTimeOffset endTimestamp)
         {
+            if (this.hasEnded)
```

Comment:
Can you clarify the use of your lock? Seems like we should be protecting all public methods that can mutate the state of the Span. Two threads could end up at line 340 simultaneously...

Also, maybe consider changing the name of @lock to something that isn't an important keyword.

Reply:
I agree the lock existence here is questionable and name could be improved. I'd prefer to address this separately from this PR - it's too big already  :) Create https://github.com/open-telemetry/opentelemetry-dotnet/issues/271 to follow up